File: rust/dijkstra.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

fn dijkstra(G: &Vec<Vec<(usize, i64)>>, dist: &mut Vec<i64>, start: usize) {
    // distの初期値はinfであることを想定
    dist[start] = 0;
    let mut hea = BinaryHeap::from(vec![(0, start)]);

    let mut visited = vec![false; G.len()];
    while let Some((_, v)) = hea.pop() {
        if visited[v] {
            continue;
        }

        visited[v] = true;
        for i in 0..G[v].len() {
            let (nv, c) = G[v][i];
            if dist[nv] > dist[v] + c {
                dist[nv] = dist[v] + c;
                // heapはmax heapであるため、負の数にする
                hea.push((-dist[nv], nv));
            }
        }
    }
}

fn dijkstra2(graph: &Vec<Vec<(i64, usize)>>, start: usize) -> Vec<i64> {
    let inf = 1e18 as i64;
    let mut dist = vec![inf; graph.len()];
    let mut seen = vec![false; graph.len()];

    let mut bheap = BinaryHeap::new();
    bheap.push((Reverse(0), start));

    dist[start] = 0;
    while let Some((Reverse(_d), v)) = bheap.pop() {
        if seen[v] {
            continue;
        }
        seen[v] = true;
        for i in 0..graph[v].len() {
            let (nd, nv) = graph[v][i];
            if dist[nv] > (dist[v] + nd) {
                dist[nv] = dist[v] + nd;
                bheap.push((Reverse(dist[nv]), nv));
            }
        }
    }

    dist
}

File: python/dijkstra.py
import heapq

def dijkstra(start: int, G: list[tuple[int]]):
    """ 経路の表現
            [終点, 辺の値]
            A, B, C, D, ... → 0, 1, 2, ...とする """
    N = len(G)
    inf = 10**18
    dist = [inf] * N    #スタート地点以外の値は∞で初期化
    dist[start] = 0     #スタートは0で初期化

    node_name = []
    heapq.heappush(node_name, [start, 0])

    while len(node_name):
        #ヒープから取り出し
        _, min_point = heapq.heappop(node_name)
        
        #経路の要素を各変数に格納することで，視覚的に見やすくする
        for nex in G[min_point]:
            goal, cost = nex

            #更新条件
            if dist[min_point] + cost < dist[goal]:
                dist[goal] = dist[min_point] + cost     #更新
                #ヒープに登録
                heapq.heappush(node_name, [dist[min_point] + cost, goal])

    return dist
